File: day12/day12.mjs
import readDataFromStdin from "../read-data.mjs";

const data = await readDataFromStdin()

function findSymbol(terrain, symbol) {
    for (let y = 0; y < terrain.length; y++) {
        for (let x = 0; x < terrain[y].length; x++) {
            if (terrain[y][x] === symbol) {
                return [x, y];
            }
        }
    }
}

function level(terrain, [x, y]) {
    const c = terrain[y][x];
    let ch
    if (c === "S") {
        ch = 'a'
    } else if (c === "E") {
        ch = 'z'
    } else {
        ch = c;
    }
    return ch.charCodeAt(0) - 96;
}

function contains(visited, [x, y]) {
    return visited[`${x}-${y}`] === true;
}

function append(visited, [x, y]) {
    visited[`${x}-${y}`] = true;
    return visited;
}

function distance([x1, y1], [x2, y2]) {
    return Math.abs(x1 - x2) + Math.abs(y1 - y2);
}

function nextDirections(terrain, [x, y], visited = {}) {
    const l = level(terrain, [x, y]);
    const directions = [[x, y - 1], [x - 1, y], [x + 1, y], [x, y + 1]]
        .filter(([x, y]) => !contains(visited, [x, y]) &&
            x >= 0 && y >= 0 && y < terrain.length && x < terrain[y].length &&
            l - level(terrain, [x, y]) <= 1
        );
    return directions;
}

function dump(terrain, visited = {}) {
    process.stdout.write("  ");
    terrain[0].forEach((_, x) => process.stdout.write((x % 10).toString()));
    process.stdout.write("\n");
    terrain.forEach((line, y) => {
        process.stdout.write(`${y % 10} `)
        line.forEach((c, x) => {
            const color = contains(visited, [x, y]);
            const last = visited.last && visited.last[0] === x && visited.last[1] === y;
            if (last) {
                process.stdout.write("\x1b[32m");
            } else if (color) {
                process.stdout.write("\x1b[31m");
            }
            process.stdout.write(c);
            if (color) {
                process.stdout.write("\x1b[0m");
            }
        });
        process.stdout.write("\n");
    });
}

const terrain = data.split("\n").map(line => line.split(""));

// dump(terrain);

// console.log('S', nextDirections(terrain, findSymbol(terrain, "S")));
// console.log('E', nextDirections(terrain, findSymbol(terrain, "E")));

function traverse(terrain, start, isEnd) {
    const queue = [];
    const visited = {};
    queue.push([start, 0]);
    append(visited, start);
    while (queue.length > 0) {
        const [current, level] = queue.shift();
        if (isEnd(current)) {
            return level;
        } else {
            for (const next of nextDirections(terrain, current, visited)) {
                append(visited, next);
                queue.push([next, level + 1]);
            };
        }
    }
    dump(terrain, visited);
    throw new Error("No path found");
}

const end = findSymbol(terrain, 'S');

const part1 = traverse(terrain, findSymbol(terrain, 'E'), current => distance(current, end) === 0);
console.log(`Part 1: ${part1}`);

const part2 = traverse(terrain, findSymbol(terrain, 'E'), current => level(terrain, current) === 1);
console.log(`Part 2: ${part2}`);

File: day12/day12.exs
defmodule Terrain do
  defstruct [:terrain, :maxX, :maxY, :levels]

  def create(terrainRaw) do
    maxX = terrainRaw |> Enum.at(0) |> String.length()
    maxY = Enum.count(terrainRaw)

    levels =
      terrainRaw
      |> Enum.with_index(fn line, y ->
        String.to_charlist(line)
        |> Enum.with_index(fn ch, x ->
          {{x, y},
           case ch do
             # 'S' -> 'a'
             83 -> 97
             # 'E' -> 'z'
             69 -> 122
             _ -> ch
           end - 97}
        end)
      end)
      |> List.flatten()
      |> Enum.into(%{})

    %Terrain{terrain: terrainRaw, maxX: maxX, maxY: maxY, levels: levels}
  end

  def findSymbol(terrain, symbol) do
    y =
      Enum.find_index(terrain.terrain, fn line ->
        String.contains?(line, symbol)
      end)

    x = String.split(Enum.at(terrain.terrain, y), symbol) |> List.first() |> String.length()
    {x, y}
  end

  def nextDirections(terrain, {x, y}) do
    l = terrain.levels[{x, y}]

    Enum.filter(
      [
        {x - 1, y},
        {x, y - 1},
        {x, y + 1},
        {x + 1, y}
      ],
      fn {x, y} ->
        x >= 0 and x < terrain.maxX and y >= 0 and y < terrain.maxY and
          l - terrain.levels[{x, y}] <= 1
      end
    )
  end

  def area(terrain), do: terrain.maxX * terrain.maxY

  def distance({x1, y1}, {x2, y2}) do
    dx = x1 - x2
    dy = y1 - y2
    trunc(:math.sqrt(:math.pow(dx, 2) + :math.pow(dy, 2)))
    # abs(dx) + abs(dy)
  end

  def dumpPath(terrain, path) do
    terrain.terrain
    |> Enum.with_index(fn line, y ->
      if y == 0 do
        tens = trunc(String.length(line) / 10)

        if tens > 0 do
          IO.write("    ")

          Enum.each(0..tens, fn col ->
            IO.write("|" <> String.pad_trailing(Integer.to_string(col * 10), 3) <> "      ")
          end)

          IO.puts("")
        end

        IO.write("    ")

        Enum.each(0..(String.length(line) - 1), fn col ->
          IO.write("#{rem(col, 10)}")
        end)

        IO.puts("")
      end

      IO.write(String.pad_leading(Integer.to_string(y), 3, " ") <> " ")

      Enum.with_index(String.codepoints(line), fn char, x ->
        color = MapSet.member?(path, {x, y})

        if color do
          IO.write("\e[31m" <> char <> "\e[0m")
        else
          IO.write(char)
        end
      end)

      IO.puts("")
    end)
  end
end

defmodule SP do
  # https://de.wikipedia.org/wiki/Breitensuche

  #   BFS(start_node, goal_node)
  #     return BFS'({start_node}, ∅, goal_node);

  #   BFS'(fringe, gesehen, goal_node)
  #     if(fringe == ∅)
  #     // Knoten nicht gefunden
  #         return false;
  #     if (goal_node ∈ fringe)
  #         return true;
  #     return BFS'({child | x ∈ fringe, child ∈ nachfolger(x)} \ gesehen, gesehen ∪ fringe, goal_node);

  def bfs(terrain, start_node, goal?) do
    bfs_run(terrain, MapSet.new([start_node]), MapSet.new(), goal?, 0)
  end

  defp bfs_run(terrain, fringe, visited, goal?, level) do
    cond do
      MapSet.size(fringe) == 0 ->
        false

      Enum.any?(fringe, goal?) ->
        level

      true ->
        new_fringe =
          fringe
          |> Enum.flat_map(fn n -> Terrain.nextDirections(terrain, n) end)
          |> Enum.filter(fn next -> not MapSet.member?(visited, next) end)
          |> Enum.into(MapSet.new())

        new_visited = MapSet.union(visited, fringe)

        bfs_run(
          terrain,
          new_fringe,
          new_visited,
          goal?,
          level + 1
        )
    end
  end
end

terrainRaw =
  IO.read(:stdio, :all)
  |> String.trim()
  |> String.split("\n")

terrain = Terrain.create(terrainRaw)

start = terrain |> Terrain.findSymbol("E")
finish = terrain |> Terrain.findSymbol("S")

SP.bfs(terrain, start, fn pos -> pos == finish end)
|> IO.inspect(label: "Part 1")

SP.bfs(terrain, start, fn pos -> terrain.levels[pos] == 0 end)
|> IO.inspect(label: "Part 2")
